lib/rufo/formatter.rb

```diff
@@ -2110,21 +2169,48 @@ def visit_array(node)
 
     _, elements = node
 
+    doc = []
+    pre_comments = []
+    post_comments = []
+    has_comment = false
     token_column = current_token_column
 
     check :on_lbracket
-    write "["
     next_token
 
     if elements
-      visit_literal_elements to_ary(elements), inside_array: true, token_column: token_column
+      doc, pre_comments, post_comments, has_comment = with_doc_mode {
+        visit_literal_elements_doc(to_ary(elements), inside_array: true, token_column: token_column)
+      }
     else
       skip_space_or_newline
     end
 
     check :on_rbracket
-    write "]"
     next_token
+    if trailing_commas && !doc.empty?
+      last = doc.pop
+      doc << B.concat([last, B.if_break(",", "")])
+    end
+
+    B.group(
```

Comment:
This expression is really nice to read 👏.

In the end, I think this points to a more readable future for rufo. Instead of needing to keep the entire stack in your head to understand a point in code, we can capture output and express the ultimate shape of an expression at the end